osu.Game/Screens/Select/BeatmapCarousel.cs

```diff
@@ -590,6 +608,10 @@ protected override void Update()
             if (revalidateItems)
                 updateYPositions();
 
+            // If an eager selection can't be made because all sets have been filtered, the flag should be reset
```

Comment:
*why* should it be reset? that seems counterintuitive. if everything has been filtered then surely that includes the "last selection", whatever it could possibly have been? (with the exception of `null`, maybe).

Reply:
Because `lastSelectionWasFiltered` is supposed to be true only until next eager selection as part of this implementation. I don't think there is a way to differentiate between user and carousel selections any other way, so that's what  `lastSelectionWasFiltered` is used for. The first selection right after filtering is the one made eagerly by the carousel.

Now that I look at it, I think I can try capturing the eager selection in `CarouselRoot.PerformSelection()` as I think that's only called eagerly. That should get rid of the need for `lastSelectionWasFiltered` and make this code way less complex.

Reply:
It looks a *little* better with the latest changes but still the change overall is more on the complex side in my view...